illumos-utils/src/running_zone.rs

```diff
@@ -137,9 +140,102 @@ pub enum GetZoneError {
+    extern "C" {
+        fn ct_tmpl_set_critical(fd: c_int, events: c_uint) -> c_int;
+        fn ct_tmpl_set_informative(fd: c_int, events: c_uint) -> c_int;
+        fn ct_pr_tmpl_set_fatal(fd: c_int, events: c_uint) -> c_int;
+        fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
+        fn ct_tmpl_activate(fd: c_int) -> c_int;
+        fn ct_tmpl_clear(fd: c_int) -> c_int;
+    }
+
+    #[link(name = "c")]
+    extern "C" {
+        pub fn zone_enter(zid: zoneid_t) -> c_int;
+    }
+
+    // A Rust wrapper around the process contract template.
+    #[derive(Debug)]
+    pub struct Template {
+        fd: c_int,
+    }
+
+    impl Drop for Template {
+        fn drop(&mut self) {
+            if unsafe { libc::close(self.fd) } != 0 {
+                let e = std::io::Error::last_os_error();
+                eprintln!("failed to close ctfs template file: {e}");
+            }
+        }
+    }
+
+    impl Template {
+        const TEMPLATE_PATH: &[u8] = b"/system/contract/process/template\0";
+
+        // Constants related to how the contract below is managed. See
+        // `usr/src/uts/common/sys/contract/process.h` in the illumos sources
+        // for details.
+
+        // Process experienced an uncorrectable error.
+        const CT_PR_EV_HWERR: c_uint = 0x20;
+        // Only kill process group on fatal errors.
+        const CT_PR_PGRPONLY: c_uint = 0x04;
+        // Automatically detach inherited contracts.
+        const CT_PR_REGENT: c_uint = 0x08;
+
+        pub fn new() -> Result<Self, crate::ExecutionError> {
+            let path = CStr::from_bytes_with_nul(Self::TEMPLATE_PATH).unwrap();
+            let fd = unsafe { libc::open(path.as_ptr(), libc::O_RDWR) };
+            if fd < 0 {
+                let err = std::io::Error::last_os_error();
+                return Err(crate::ExecutionError::ZoneEnter { err });
+            }
+
+            // Initialize the contract template.
+            //
+            // No events are delivered, nothing is inherited, and we do not
+            // allow the contract to be orphaned.
+            //
+            // See illumos sources in `usr/src/cmd/zlogin/zlogin.c` in the
+            // implementation of `init_template()` for details.
+            if unsafe { ct_tmpl_set_critical(fd, 0) } != 0
+                || unsafe { ct_tmpl_set_informative(fd, 0) } != 0
+                || unsafe { ct_pr_tmpl_set_fatal(fd, Self::CT_PR_EV_HWERR) }
+                    != 0
+                || unsafe {
+                    ct_pr_tmpl_set_param(
+                        fd,
+                        Self::CT_PR_PGRPONLY | Self::CT_PR_REGENT,
+                    )
+                } != 0
+                || unsafe { ct_tmpl_activate(fd) } != 0
+            {
+                let err = std::io::Error::last_os_error();
+                return Err(crate::ExecutionError::ZoneEnter { err });
+            }
+            Ok(Self { fd })
+        }
+
+        pub fn clear(&self) {
```

Comment:
Should this return an error on failure?

Reply:
Maybe? I'm not sure what to do with it though. In particular, suppose that we successfully execute the subprocess in the zone, but calling `clear()` fails. Should that return an error, even though the actual guts of `run_cmd()` worked?